firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalDocumentsView.java

```diff
@@ -325,27 +351,65 @@ void recalculateAndSaveOverlays(Set<DocumentKey> documentKeys) {
 
     // As documents might match the query because of their overlay we need to include all documents
     // in the result.
+    remoteDocuments = updateRemoteDocumentsWithOverlayDocuments(remoteDocuments, overlays.keySet());
+
+    // Apply the overlays and match against the query.
+    ImmutableSortedMap<DocumentKey, Document> results = emptyDocumentMap();
+    for (Map.Entry<DocumentKey, MutableDocument> docEntry : remoteDocuments) {
+      Mutation overlay = overlays.get(docEntry.getKey());
+      if (overlay != null) {
+        overlay.applyToLocalView(docEntry.getValue(), null, Timestamp.now());
+      }
+      // Finally, insert the documents that still match the query
+      if (query.matches(docEntry.getValue())) {
+        results = results.insert(docEntry.getKey(), docEntry.getValue());
+      }
+    }
+
+    return results;
+  }
+
+  /** Updates the provided remote documents to include documents in the overlay. */
+  private ImmutableSortedMap<DocumentKey, MutableDocument>
```

Comment:
I tried refactoring out shared code between `getDocumentsMatchingCollectionQueryFromOverlayCache` and `getDocumentsCollectionQueryWithBatchId`, but I think it can probably be cleaner.

Reply:
We are probably at a point where we should just remove the non-overlay code.

@wu-hui